contracts/tokenomics/generator/src/contract.rs

```diff
@@ -323,61 +323,31 @@ pub fn execute_update_config(
 ///
 /// * **env** is an object of type [`Env`].
 ///
-/// * **lp_token** is an object of type [`Addr`]. This is the LP token for which to create a generator.
-///
-/// * **alloc_point** is an object of type [`Uint64`]. This is the portion of ASTRO emissions that the generator gets.
-///
-/// * **has_asset_rewards** is the field of type [`bool`]. This flag indicates whether the generator receives dual rewards.
-///
-/// * **reward_proxy** is an [`Option`] field object of type [`String`]. This is the address of the dual rewards proxy.
+/// * **pools** is a vector of set that contains LP token address and allocation point.
 ///
 /// ##Executor
-/// Can only be called by the owner.
-pub fn add(
+/// Can only be called by the owner or generator controller
+pub fn setup_pools(
     mut deps: DepsMut,
     env: Env,
-    lp_token: Addr,
-    alloc_point: Uint64,
-    has_asset_rewards: bool,
-    reward_proxy: Option<String>,
+    pools: Vec<(Addr, Uint64)>,
 ) -> Result<Response, ContractError> {
     let mut cfg = CONFIG.load(deps.storage)?;
 
-    if POOL_INFO.load(deps.storage, &lp_token).is_ok() {
-        return Err(ContractError::TokenPoolAlreadyExists {});
-    }
-
-    let reward_proxy = reward_proxy
-        .map(|v| addr_validate_to_lower(deps.api, &v))
-        .transpose()?;
+    mass_update_pools(deps.branch(), env.clone())?;
 
-    if let Some(proxy) = &reward_proxy {
-        if !cfg.allowed_reward_proxies.contains(proxy) {
-            return Err(ContractError::RewardProxyNotAllowed {});
+    for (lp_token, _) in pools.iter() {
+        if POOL_INFO.may_load(deps.storage, lp_token)?.is_none() {
```

Comment:
let's use `POOL_INFO.has` method:
/// has returns true or false if any data is at this key, without parsing or interpreting the

Reply:
may_load is necessary to raise deserialization errors